In [6]:
# code created by Rishabh Sinha
def discount_rate(x, t):
    return x**(1/t) - 1

print(f"x = {4}, t = {19}, discount rate: {discount_rate(4, 19):.4f} = {discount_rate(4, 19)*100:.2f}%")

k, x, t = 1000, 4, 19
print(f"Initial Sum ${k} -> ${k * (1 + rate)**t:.2f} in {t} years")
print(f"Expected Growth: ${k * x:.2f}")

x = 4, t = 19, discount rate: 0.0757 = 7.57%
Initial Sum $1000 -> $4000.00 in 19 years
Expected Growth: $4000.00


1000(1 + 0.0757)^19 = 4000.6652
A 4x increase after 19 years is 1000 x 4 = 4000.
Thus verified by hand calculations

In [13]:
# code created by Rishabh Sinha
import pandas as pd
import numpy as np

data = {
    't': ['1/1/2023', '3/1/2023', '4/1/2023', '6/1/2023', 
             '9/1/2023', '10/1/2023', '11/1/2023', '12/31/2023'],
    'B_t': [200000, 197500, 212000, 216000, 215000, 221500, 224000, None],
    'W_t': [0, 10000, -2000, -1500, 4000, 3500, 1500, 0],
    'B_t_next_period': [200000, 207500, 210000, 214500, 219000, 225000, 225500, 231000]
}

df = pd.DataFrame(data)
df.loc[7, 'B_t'] = df.loc[7, 'B_t_next_period']
df['i_t'] = np.nan

for i in range(1, len(df)):
    df.loc[i, 'i_t'] = (df.loc[i, 'B_t'] / (df.loc[i-1, 'B_t'] + df.loc[i-1, 'W_t'])) - 1
    
print(df[['t', 'B_t', 'W_t', 'i_t']])
print(f"\nTime-weighted rate of return for 2023: {(np.prod(1 + df['i_t'][1:]) - 1):.4f} or {(np.prod(1 + df['i_t'][1:]) - 1)*100:.2f}%")

            t       B_t    W_t       i_t
0    1/1/2023  200000.0      0       NaN
1    3/1/2023  197500.0  10000 -0.012500
2    4/1/2023  212000.0  -2000  0.021687
3    6/1/2023  216000.0  -1500  0.028571
4    9/1/2023  215000.0   4000  0.002331
5   10/1/2023  221500.0   3500  0.011416
6   11/1/2023  224000.0   1500 -0.004444
7  12/31/2023  231000.0      0  0.024390

Time-weighted rate of return for 2023: 0.0729 or 7.29%
